### ISSUE TITLE
Not OK: perl v5.7.0 +DEVEL7523 on os2-64int-ld 2.30 (UNINSTALLED)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2801]
Migrated from [rt.perl.org#4586](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4586) (status was 'resolved')

Searchable as RT4586$

**Comment by p5pRT** [comment]
### From @ysth



comp/package\.\.\.\.\.\.\.\.\.ok
comp/proto\.\.\.\.\.\.\.\.\.\.\.Scalars leaked&#8203;: 1
ok
comp/redef\.\.\.\.\.\.\.\.\.\.\.ok
\.\.\.
pragma/warnings\.\.\.\.\.\.PROG&#8203;: 
\# utf8\.c \[utf8\_to\_uv\] \-W
use utf8 ;
my $a = "snxstorm" ;
\{
&nbsp;    no warnings 'utf8' ;
&nbsp;    my $a = "snxstorm";
&nbsp;    use warnings 'utf8' ;
&nbsp;    my $a = "snxstorm";
\}
EXPECTED&#8203;:
Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 3\.
Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 8\.
GOT&#8203;:
Malformed UTF\-8 character \(unexpected non\-continuation byte 0xf800000073 after byte 0x4606e00000000\) at \- line 3\.
Malformed UTF\-8 character \(unexpected non\-continuation byte 0xf800000073 after byte 0x4606e00000000\) at \- line 8\.

Process terminated by SIGPIPE
FAILED at test 410
lib/abbrev\.\.\.\.\.\.\.\.\.\.\.ok

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=none

Site configuration information for perl v5.7.0:

Configured by sthoenna at Fri Nov  3 14:53:11 PST 2000.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=os2, osvers=2.30, archname=os2-64int-ld
    uname='os2 efn.org 2 2.30 i386  '
    config_args='-de -Dprefix=d:/perl -Dusedevel -Duse64bitint -Duselongdouble -Aoptimize=-DDEBUGGING'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
  Compiler:
    cc='gcc', ccflags ='-Zomf -Zmt -DDOSISH -DOS2=2 -DEMBED -I. -D_EMX_CRT_REV_=63',
    optimize='-O2 -fomit-frame-pointer -malign-loops=2 -malign-jumps=2 -malign-functions=2 -s -DDEBUGGING',
    cppflags='-Zomf -Zmt -DDOSISH -DOS2=2 -DEMBED -I. -D_EMX_CRT_REV_=63'
    ccversion='', gccversion='2.8.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-Zexe -Zomf -Zmt -Zcrtdll -Zstack 32000'
    libpth=d:/emx/lib d:/emx/lib/mt
    libs=-lsocket -lm -lbsd
    perllibs=-lsocket -lm -lbsd
    libc=d:/emx/lib/mt/c_import.lib, so=dll, useshrplib=true, libperl=libperl.lib
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-Zdll', lddlflags='-Zdll -Zomf -Zmt -Zcrtdll -s'

Locally applied patches:
    DEVEL7523


@INC for perl v5.7.0:
    lib
    d:/perl/lib/5.7.0/os2-64int-ld
    d:/perl/lib/5.7.0
    d:/perl/lib/site_perl/5.7.0/os2-64int-ld
    d:/perl/lib/site_perl/5.7.0
    d:/perl/lib/site_perl
    .


Environment for perl v5.7.0:
    HOME=d:\home\sthoenna
    LANG=en_us
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/perl-current
    LOGDIR=sthoenna
    PATH=d:\bin;C:\OS2;d:\perl\bin;C:\OS2\SYSTEM;C:\OS2\INSTALL;C:\;C:\OS2\MDOS;C:\OS2\APPS;C:\MMOS2;d:\os2apps\util;d:\DOSAPPS\UTIL;c:\sio;D:\WINDOWS;d:\pdksh;d:\emx\bin;d:\emacs\19.33\bin;d:\ispell
    PERL_BADLANG (unset)
    PERL_SH_DIR=d:\BIN
    SHELL=d:/bin/sh.exe



```
</details>

**Comment by p5pRT** [comment]
### From @ysth

Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> On Fri\, Nov 03\, 2000 at 03&#8203;:31&#8203;:42PM \-0800\, sthoenna@&#8203;efn\.org wrote&#8203;:
> > This is a build failure report for perl from sthoenna@&#8203;efn\.org\,
> > generated with the help of perlbug 1\.32 running under perl v5\.7\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > comp/package\.\.\.\.\.\.\.\.\.ok
> > comp/proto\.\.\.\.\.\.\.\.\.\.\.Scalars leaked&#8203;: 1
> > ok
> > comp/redef\.\.\.\.\.\.\.\.\.\.\.ok
> > \.\.\.
> > pragma/warnings\.\.\.\.\.\.PROG&#8203;: 
> > \# utf8\.c \[utf8\_to\_uv\] \-W
> > use utf8 ;
> > my $a = "snxstorm" ;
> > \{
> >     no warnings 'utf8' ;
> >     my $a = "snxstorm";
> >     use warnings 'utf8' ;
> >     my $a = "snxstorm";
> > \}
> > EXPECTED&#8203;:
> > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 3\.
> > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 8\.
> > GOT&#8203;:
> > Malformed UTF\-8 character \(unexpected non\-continuation byte 0xf800000073 after byte 0x4606e00000000\) at \- line 3\.
> 
> Looks like a combined sign extension and 64\-bitness problem\.  What are your
> U8\, UV\, and UVxf?

\[D&#8203;:\\perl\-current\]fgrep \-we UVTYPE \-e U8TYPE \-e UVxf config\.h|fgrep define
\#define UVTYPE          unsigned long long              /\*\*/
\#define U8TYPE          unsigned char           /\*\*/
\#define UVxf            "llx"           /\*\*/

**Comment by p5pRT** [comment]
### From @jhi

> > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 3\.
> > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 8\.
> > > GOT&#8203;:
> > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0xf800000073 after byte 0x4606e00000000\) at \- line 3\.
> > 
> > Looks like a combined sign extension and 64\-bitness problem\.  What are your
> > U8\, UV\, and UVxf?
> 
> \[D&#8203;:\\perl\-current\]fgrep \-we UVTYPE \-e U8TYPE \-e UVxf config\.h|fgrep define
> \#define UVTYPE          unsigned long long              /\*\*/
> \#define U8TYPE          unsigned char           /\*\*/
> \#define UVxf            "llx"           /\*\*/

That looks okay\.  Now\, the question is\, why did it not work? :\-\)
The error message comes from utf8\.c&#8203;:utf8\_to\_uv\(\)\.  Try

\#include \<stdio\.h>
int main\(\) \{
unsigned char \*s = "\\xf8";
unsigned long long c = \*s;
printf\("%llx\\n"\, c\);
return 0;
\}

because this is what fails&#8203;:

&nbsp;                        "Malformed UTF\-8 character \(unexpected non\-continuation byte 0x%02"UVxf" after byte 0x%02"UVxf"\)"\,
&nbsp;                        s\[1\]\, uv\);

where

UV
Perl\_utf8\_to\_uv\(pTHX\_ U8\* s\, STRLEN curlen\, STRLEN\* retlen\, U32 flags\)
\{
&nbsp;    dTHR;
&nbsp;    UV uv = \*s\, ouv;

In other words the first format is

&nbsp;    "%02llx" for a U8 argument

and the second one

&nbsp;    "%02llx" for a UV argument

So I'm rather confused of what could be going wrong\.  I suggest turning
up the warnings of the compiler\.

**Comment by p5pRT** [comment]
### From @jhi

> Test program prints "f8\\n"\.

Rats\.  There goes another fine theory\.

> I'll look into it further when I get a chance\.  Any ideas about the
> comp/proto scalar leak?  \(New since 7483\)\.

Nope\.  If someone could binary search which of the changes contained
in 7483 is the culprit\, that would be nice \(there are actually only
a few code changes in there\)\.

**Comment by p5pRT** [comment]
### From @ysth

Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> > > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 3\.
> > > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0x73 after byte 0xf8\) at \- line 8\.
> > > > GOT&#8203;:
> > > > Malformed UTF\-8 character \(unexpected non\-continuation byte 0xf800000073 after byte 0x4606e00000000\) at \- line 3\.
> > > 
> > > Looks like a combined sign extension and 64\-bitness problem\.  What are your
> > > U8\, UV\, and UVxf?
> > 
> > \[D&#8203;:\\perl\-current\]fgrep \-we UVTYPE \-e U8TYPE \-e UVxf config\.h|fgrep define
> > \#define UVTYPE          unsigned long long              /\*\*/
> > \#define U8TYPE          unsigned char           /\*\*/
> > \#define UVxf            "llx"           /\*\*/
> 
> That looks okay\.  Now\, the question is\, why did it not work? :\-\)
> The error message comes from utf8\.c&#8203;:utf8\_to\_uv\(\)\.  Try
> 
> \#include \<stdio\.h>
> int main\(\) \{
> unsigned char \*s = "\\xf8";
> unsigned long long c = \*s;
> printf\("%llx\\n"\, c\);
> return 0;
> \}
> 
> because this is what fails&#8203;:
> 
>                         "Malformed UTF\-8 character \(unexpected non\-continuation byte 0x%02"UVxf" after byte 0x%02"UVxf"\)"\,
>                         s\[1\]\, uv\);
> 
> where
> 
> UV
> Perl\_utf8\_to\_uv\(pTHX\_ U8\* s\, STRLEN curlen\, STRLEN\* retlen\, U32 flags\)
> \{
>     dTHR;
>     UV uv = \*s\, ouv;
> 
> In other words the first format is
> 
>     "%02llx" for a U8 argument
> 
> and the second one
> 
>     "%02llx" for a UV argument
> 
> So I'm rather confused of what could be going wrong\.  I suggest turning
> up the warnings of the compiler\.

test program prints "f8\\n"\.  I'll look into it further when I get a
chance\.  Any ideas about the comp/proto scalar leak?  \(New since 7483\)\.

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Sun\, 5 Nov 2000 15&#8203;:30&#8203;:32 \-0600\, Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> said&#8203;:

>> I'll look into it further when I get a chance\.  Any ideas about the
>> comp/proto scalar leak?  \(New since 7483\)\.

&nbsp;      > Nope\.  If someone could binary search which of the changes contained
&nbsp;      > in 7483 is the culprit\, that would be nice \(there are actually only
&nbsp;      > a few code changes in there\)\.

The proto scalar leak comes from my tests that I added at the end of
comp/proto\.t\. They contain syntax errors caught by eval\.

**Comment by p5pRT** [comment]
### From @jhi

On Mon\, Nov 06\, 2000 at 12&#8203;:23&#8203;:52AM \+0100\, Andreas J\. Koenig wrote&#8203;:
> >>>>> On Sun\, 5 Nov 2000 15&#8203;:30&#8203;:32 \-0600\, Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> said&#8203;:
> 
> >> I'll look into it further when I get a chance\.  Any ideas about the
> >> comp/proto scalar leak?  \(New since 7483\)\.
> 
>       > Nope\.  If someone could binary search which of the changes contained
>       > in 7483 is the culprit\, that would be nice \(there are actually only
>       > a few code changes in there\)\.
> 
> The proto scalar leak comes from my tests that I added at the end of
> comp/proto\.t\. They contain syntax errors caught by eval\.

Ahhh\.  I'll document that to comp/proto\.t\.

> \-\- 
> andreas

**Comment by p5pRT** [comment]
### From @tux

On 06 Nov 2000 00&#8203;:23&#8203;:52 \+0100\, andreas\.koenig@&#8203;anima\.de \(Andreas J\. Koenig\) wrote&#8203;:
> >>>>> On Sun\, 5 Nov 2000 15&#8203;:30&#8203;:32 \-0600\, Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> said&#8203;:
> 
> >> I'll look into it further when I get a chance\.  Any ideas about the
> >> comp/proto scalar leak?  \(New since 7483\)\.
> 
>       > Nope\.  If someone could binary search which of the changes contained
>       > in 7483 is the culprit\, that would be nice \(there are actually only
>       > a few code changes in there\)\.
> 
> The proto scalar leak comes from my tests that I added at the end of
> comp/proto\.t\. They contain syntax errors caught by eval\.

Are they releated to the huge list from tie\-hashref posted on 2 Nov?

perl@&#8203;7528 HP\-UX 11\.00

Though all test succeed\, this doesn't feel right&#8203;:

\# perl TEST lib/tie\-refhash\.t
lib/tie\-refhash\.\.\.Unbalanced string table refcount&#8203;: \(1\) for "dumph" during global destruction\.
Unbalanced string table refcount&#8203;: \(1\) for "STDIN" during global destruction\.
Unbalanced string table refcount&#8203;: \(1\) for " during global destruction\.
Unbalanced string table refcount&#8203;: \(1\) for "\_\<\.\./lib/overload\.pm" during global destruction\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

H \. Merijn Brand \<h\.m\.brand@&#8203;hccnet\.nl> writes&#8203;:
>On 06 Nov 2000 00&#8203;:23&#8203;:52 \+0100\, andreas\.koenig@&#8203;anima\.de \(Andreas J\. Koenig\) wrote&#8203;:
>> >>>>> On Sun\, 5 Nov 2000 15&#8203;:30&#8203;:32 \-0600\, Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> said&#8203;:
>> 
>> >> I'll look into it further when I get a chance\.  Any ideas about the
>> >> comp/proto scalar leak?  \(New since 7483\)\.
>> 
>>       > Nope\.  If someone could binary search which of the changes contained
>>       > in 7483 is the culprit\, that would be nice \(there are actually only
>>       > a few code changes in there\)\.
>> 
>> The proto scalar leak comes from my tests that I added at the end of
>> comp/proto\.t\. They contain syntax errors caught by eval\.
>
>Are they releated to the huge list from tie\-hashref posted on 2 Nov?
>
>perl@&#8203;7528 HP\-UX 11\.00
>
>Though all test succeed\, this doesn't feel right&#8203;:
>
>\# perl TEST lib/tie\-refhash\.t
>lib/tie\-refhash\.\.\.Unbalanced string table refcount&#8203;: \(1\) for "dumph" during global destruction\.
>Unbalanced string table refcount&#8203;: \(1\) for "STDIN" during global destruction\.
>Unbalanced string table refcount&#8203;: \(1\) for " during global destruction\.

FWIW \- the keys that have not been decremented seem to be those of 
%main&#8203;:: i\.e\. the top level stash\.

The main problem in tracking it down is lib/tie\-refhash\.t is one 
of the most complex pieces of perl code I have seen\. With nested eval ""\,
eval \{\} and Data&#8203;::Dumper involved as well as Tie&#8203;::RefHash\. The evals 
are catching things as well\. 
&nbsp;

**Comment by p5pRT** [comment]
### From @ysth

Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> > Test program prints "f8\\n"\.
> 
> Rats\.  There goes another fine theory\.

/\*&#8203;:\( \(smacks self on head\)

var args don't work to well if you put an unsigned char on the stack
\(which seems to use 4 bytes on my implementation\) and pop an unsigned
long long off\.  Unless you have a real 64bit machine\.

<details><summary>Inline Patch</summary>

```diff
--- utf8.c#7573	Thu Nov  2 22:47:20 2000
+++ utf8.c	Mon Nov  6 14:54:16 2000
@@ -215,7 +215,7 @@
 	if (dowarn)
 	    Perl_warner(aTHX_ WARN_UTF8,
 			"Malformed UTF-8 character (unexpected non-continuation byte 0x%02"UVxf" after byte 0x%02"UVxf")",
-			s[1], uv);
+			(UV)s[1], uv);
 	goto malformed;
     }
     

```
</details>
End of Patch.

> > I'll look into it further when I get a chance\.  Any ideas about the
> > comp/proto scalar leak?  \(New since 7483\)\.
> 
> Nope\.  If someone could binary search which of the changes contained
> in 7483 is the culprit\, that would be nice \(there are actually only
> a few code changes in there\)\.

I meant it was ok in 7483\.  The problem appeared in \(7483\, 7524\]\.
Still there in 7573\.
